src/app/qgsmaptooladdring.cpp

```diff
@@ -29,6 +29,8 @@ QgsMapToolAddRing::QgsMapToolAddRing( QgsMapCanvas* canvas )
     : QgsMapToolCapture( canvas, QgisApp::instance()->cadDockWidget(), QgsMapToolCapture::CapturePolygon )
 {
   mToolName = tr( "Add ring" );
+  connect( QgisApp::instance(), SIGNAL( newProject() ), this, SLOT( stopCapturing() ) );
+  connect( QgisApp::instance(), SIGNAL( projectRead() ), this, SLOT( stopCapturing() ) );
```

Comment:
I agree it shouldn't persist, but it always did and it does at the moment. 
* Create layer A
* Start capturing
* Remove layer A
* Add layer B
* Continue and finish capturing
* Everything is fine

While I think binding the capture process to a layer is a good idea, it is more than I can do for 2.14.